### PR TITLE
DM-33278: Improve storage class conversion

### PR DESCRIPTION
**Pull request by timj**
* Add a dataset type compatibility method.
* Add support for specifying methods.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: doc/lsst.daf.butler/formatters.rst [hunk 1]
```diff
@@ -139,6 +146,22 @@ You can do this using YAML anchors and references but the preferred approach is
...
+      pytype: dict
+      converters:
+        lsst.daf.base.PropertySet: toDict()
+    TaskMetadata:
+      pytype: lsst.pipe.base.TaskMetadata
+      converters:
+        lsst.daf.base.PropertySet: lsst.pipe.base.TaskMetadata.from_metadata
+
+In the first definition, the configuration says that if a ``PropertySet`` object is given then the ``toDict()`` method can be called on it and the returned value will be a `dict`.
+In the second definition a ``PropertySet`` is again specified but this time the ``from_metadata`` class method will be called with the ``PropertySet`` as the first parameter and a ``TaskMetadata`` will be returned.
```
Comment (andy-slac) [inline doc/lsst.daf.butler/formatters.rst]: This difference between `()` and no `()` confused me a lot when I first saw it and it was not obvious what it means. I had to read everything more than once to figure it out. Now I think that `toDict()` is basically the same as `lsst.daf.base.PropertySet.toDict` (non-bound method). Would it be better to drop `()` and use non-bound method syntax for symmetry?

File: python/lsst/daf/butler/core/datasets/type.py [hunk 2]
```diff
@@ -240,25 +240,78 @@ def __repr__(self) -> str:
...
         if self._storageClass is not None and other._storageClass is not None:
             if self._storageClass != other._storageClass:
                 return False
         else:
             if self._storageClassName != other._storageClassName:
                 return False
-        if self._isCalibration != other._isCalibration:
+        return True
+
+    def is_compatible_with(self, other: "DatasetType") -> bool:
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/datasets/type.py]: Do you need to quote `DatasetType` here?

Comment (andy-slac) [inline python/lsst/daf/butler/core/datasets/type.py]: Does "compatible" mean compatible for both `get` and `put`? Is it symmetrical (X.is_compatible_with(Y) === Y.is_compatible_with(X))?

**Review by andy-slac — APPROVED** [review]
Looks good, couple of minor comments/questions.

File: python/lsst/daf/butler/core/datasets/type.py (hunk 2 above)
Comment (timj) [reply python/lsst/daf/butler/core/datasets/type.py]: It really means `self_sc.can_convert(other_sc)` -- You have made me realize that ctrl_mpexec is going to have to try to distinguish inputs (so can convert on get to the required type for the Task) and outputs (so can convert the type used in the Task to the required type in the butler).

Comment (timj) [reply python/lsst/daf/butler/core/datasets/type.py]: I think I'm confused between `from __future__ import annotations` being enabled and not being enabled. There was some other place where we did have to use strings if the type was used in the context of its own class but maybe that was to do with pydantic.

Comment (TallJimbo) [reply python/lsst/daf/butler/core/datasets/type.py]: I believe we cannot use `from __future__ import annotations` with pydantic (which is really a shame, and I think it means we should mostly put pydantic classes in their own modules, so we can use it everywhere else).  And without `__future__.annotations`, you do have to quote in places like this.

File: doc/lsst.daf.butler/formatters.rst (hunk 1 above)
Comment (timj) [reply doc/lsst.daf.butler/formatters.rst]: I've been pondering this and the `()` syntax has the advantage that it doesn't lead to duplicating the full class name on both sides:

```yaml
lsst.daf.butler.tests.MetricsExample: exportAsDict()
```
vs
```yaml
lsst.daf.butler.tests.MetricsExample: lsst.daf.butler.tests.MetricsExample.exportAsDict
```
The latter does work and doesn't require any code changes to make it work without this ticket. Maybe that's a good enough argument for not trying to bother with the more succinct form.

**Pull request merged**